### PR TITLE
ROT47 decoder

### PR DESCRIPTION
**Pull request by aldogutierrez**
Finished ROT47 decoder.

---------------------
I made the mistake to commit and push without making a new branch, so I reverted those changes and pushed them to a new branch, I apologize for any inconvenience

### CODE REVIEW

**Comment by bee-san** [comment]
Hey, unfortunately ROT47 is a cracker, not a decoder so we can only accept submissions in the form of C++ code. We already have a ROT26 cracker, so to make it ROT47 I am under the assumption that we only need to change the alphabet out.

**Comment by Cyclic3** [comment]
No it's not, this is a decoder

**Comment by Cyclic3** [comment]
honestly

**Comment by bee-san** [comment]
Hey, please fix this:

```
       for character in encrypted_words:
>           decoded_words += decoder_dict.get(character)
E           TypeError: can only concatenate str (not "NoneType") to str
```

You can test it is fixed by running `pytest` on your machine :)

File: ciphey/basemods/Decoders/rot47.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        # Uniting both alphabets, we get the decoding key
+        decoder_dict = dict(zip(encode_alphabet, decoder_alphabet))
+
+        # From all the words of the ciphertext, we tokenize every letter for easy retrieval from the dictionary
+        encrypted_words = list(ctext)
+        # Empty string in which the plaintext will be concatenated
+        decoded_words = ''
+
+        # We get the values from the dictionary, and build back the plaintext
+        for character in encrypted_words:
```
Comment (bee-san) [inline ciphey/basemods/Decoders/rot47.py:32]: this is what is breaking :)

File: ciphey/basemods/Decoders/rot47.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    """
+    def decode(self, ctext: T) -> Optional[U]:
+        """
+        The decoder function for the ROT47 algorithm
+
+        :param ctext: The input ciphertext
+        :return: The decoded string from the ciphertext
+        """
+        # The ciphertext will be encoded using a ROT47 alphabet, here, named encode_alphabet
+        encode_alphabet = '!\"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~ '
```
Comment (bee-san) [inline ciphey/basemods/Decoders/rot47.py:18]: This should go into CipheyDists, see the wiki for help :)

File: ciphey/basemods/Decoders/rot47.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+        # The decoder_alphabet can reserve the cipher and turn it into the plaintext
+        decoder_alphabet = 'PQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~!\"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNO '
```
Comment (bee-san) [inline ciphey/basemods/Decoders/rot47.py:21]: This should go into CipheyDists, see the wiki for help :)

File: ciphey/basemods/Decoders/rot47.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+        # Uniting both alphabets, we get the decoding key
+        decoder_dict = dict(zip(encode_alphabet, decoder_alphabet))
```
Comment (bee-san) [inline ciphey/basemods/Decoders/rot47.py:24]: This should go into CipheyDists, see the wiki for help :)

File: ciphey/basemods/Decoders/rot47.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        # Return the plaintext
+        return decoded_words
+
+    @staticmethod
+    def priority() -> float:
+        """
+        Set to pretty low
+        :return:
+        """
+        return 0.5
```
Comment (bee-san) [inline ciphey/basemods/Decoders/rot47.py:44]: Priority should be way lower, try 0.05 :)

**Review by bee-san — CHANGES_REQUESTED** [review]

**Comment by SkeletalDemise** [comment]
I'm closing this because ROT47 will actually need to be a C++ cracker. ROT47 uses an alphabet of ASCII characters 33 to 126, making an alphabet length of 94. Only adding ROT47 would be like only adding ROT13 without adding ROT 1-25. 

We'll need a cracker that can decode ROT 1-94 using the ROT47 alphabet. It should be easy to just use the existing caesar cracker with the ROT47 alphabet.

**Pull request closed without merging**